src/components/file-input/file-input.stories.tsx

```diff
@@ -0,0 +1,320 @@
+  return <FileInput label="File input" />;
+};
+
+export const Margins: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" m={5} />;
+};
+
+export const Disabled: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" disabled />;
+};
+
+export const WithInputHint: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" inputHint="Hint text" />;
+};
+
+export const Required: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" required />;
+};
+
+export const IncreasedHeight: ComponentStory<typeof FileInput> = () => {
+  return (
+    <FileInput
+      label="File input"
+      dragAndDropText="You can drag and drop your file here, if that's the way you prefer to interact with the component."
+      maxHeight="200px"
+    />
+  );
+};
+
+export const IncreasedWidth: ComponentStory<typeof FileInput> = () => {
+  return (
+    <FileInput
+      label="File input"
+      dragAndDropText="You can drag and drop your file here, if that's the way you prefer to interact with the component."
+      maxWidth="750px"
+    />
+  );
+};
+
+export const IncreasedBoth: ComponentStory<typeof FileInput> = () => {
+  return (
+    <FileInput
+      label="File input"
+      dragAndDropText="You can drag and drop your file here, if that's the way you prefer to interact with the component."
+      maxWidth="500px"
+      maxHeight="200px"
+    />
+  );
+};
+
+export const FullWidth: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" maxWidth="100%" />;
+};
+
+export const Vertical: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="File input" isVertical />;
+};
+
+export const Validation: ComponentStory<typeof FileInput> = () => {
+  return (
+    <>
+      <FileInput
+        m={4}
+        label="error as string"
+        inputHint="Hint text (optional)"
+        error="error message"
+      />
+      <FileInput
+        m={4}
+        label="error as boolean"
+        inputHint="Hint text (optional)"
+        error
+      />
+    </>
+  );
+};
+
+export const Accept: ComponentStory<typeof FileInput> = () => {
+  return <FileInput label="Only accepts image files" accept="image/*" />;
+};
+
+export const FileTypeValidation: ComponentStory<typeof FileInput> = () => {
+  const [error, setError] = useState<string | undefined>();
+  const onChange = (files: FileList) => {
+    let errorMessage;
+    if (files.length > 0) {
+      const fileType = files[0].type;
+      if (!fileType.startsWith("image/")) {
+        errorMessage = "Please choose an image file to upload";
+      }
+    }
+    setError(errorMessage);
+  };
+  return (
+    <FileInput
+      label="Only accepts image files"
+      accept="image/*"
+      error={error}
+      onChange={onChange}
+    />
+  );
+};
+
+export const UploadStatusClient: ComponentStory<typeof FileInput> = () => {
```

Comment:
comment: I've been going backwards and forwards on this a bit so happy for this to be a bit more of an open discussion around it so we can decide etc. I'm leaning towards this being baked into the component as most use cases will need this and adding an override to allow for any cases that don't.

As I say happy for to open this into a wider discussion as there's potentially scenarios I'm missing etc. I had a little go at moving it into the component and fixing the linting complaints but it may be there's a better way of doing it 🤷 (if we all agree it should be baked in obviously)

https://gist.github.com/edleeks87/a3968fbbd2684868150d5a29e0d5ef5a

Reply:
I definitely think it's worth a discussion, but I am wondering whether leaving out this additional complexity is the better approach - as it's an extra aspect to maintain beyond primarily just the UI of it. Thinking about this some more I think we did say that the implementation detail should be down to whichever product is consuming it, and I think offering an example of how it could be used (as Robin has done nicely), should be enough.

But yeah, let's get something in the calendar though as I'm not beholden to that - and it would be good to shore up our approach to handling implementation detail like this in the future.

Reply:
Happy to put a meeting in for next week on this - I agree it's a tricky question.

For what it's worth I started off wanting to have this baked in as a default, that consumers are allowed to override - but as I started to implement it the level of complexity became such that I really wasn't comfortable putting so much code in basically a big `if` statement (it would have ended up being more than half the component's code I think!), especially as I suspect there would have been other complications. And I'm not even sure that many consumers will want to do this anyway (I really have no idea though), so is there massive benefit? Documenting it in Storybook seemed - and still seems - to be the best compromise to me.

Regarding the eslint-disable - if you know a way to get around that @edleeks87 please let me know! I soon convinced myself it wasn't possible because there are functions that need to reference each other (not by direct recursion, but by adding/removing event listeners) so that whichever order they're defined in, eslint isn't happy. I don't think it's a big deal here to be honest but if there is a way to keep eslint happy without disabling it and without introducing more problems I'd be very happy to put that in!

Reply:
actually @edleeks87 I only just looked at your gist and it seems pretty much OK 👍 (Not that I've tested it yet, but I'd be surprised if there were any major problems.) The only thing is it will still run all this code if consumers provide their own `onChange`, which it shouldn't - however now I think about it, just adding an `if (onChange) return;` early in `onInputChange` should be all it takes to do that. I'm now coming round more to your point of view and am in 2 minds - going to put in a short meeting to discuss this.